content/interoperability.tex

```diff
@@ -0,0 +1,191 @@
+and performance of hybrid programs, and definition of an OpenSHMEM
+API that queries the interoperability features provided by an \openshmem library.
+
+
+\section{MPI Interoperability}
+
+\openshmem and MPI are two commonly used parallel programming models for
+distributed-memory systems. The user can choose to utilize both models in the same program
+to efficiently and easily support various communication patterns.
+
+A vendor may implement the \openshmem and MPI libraries in different ways. For
+instance, one may implement both \openshmem and MPI as standalone libraries,
+each of which allocates and initializes fully isolated communication
+resources.
+Another common approach
+is to implement both \openshmem and MPI interfaces within the
+same software system in order to share a communication resource when possible.
+
+To improve interoperability and portability in \openshmem + MPI hybrid
+programming, we clarify the relevant semantics in the following subsections.
+
+
+\subsection{Initialization}
+In order to ensure that a hybrid program can be portably performed with different vendor
+implementations, the \openshmem environment of the program must be initialized by
+a call to \FUNC{shmem\_init} or \FUNC{shmem\_init\_thread} and be finalized by
+a call to \FUNC{shmem\_finalize}; the MPI environment of the program must be initialized
+by a call to \FUNC{MPI\_Init} or \FUNC{MPI\_Init\_thread} and be finalized by a
+call to \FUNC{MPI\_Finalize}.
+
+\apiimpnotes{
+Portable implementations of OpenSHMEM and MPI must ensure that the initialization
+calls can be made in an arbitrary order within a program; the same rule also
+applies to the finalization calls. A software runtime that utilizes a shared
+communication resource for \openshmem and MPI communication may maintain an
+internal reference counter in order to ensure that the shared resource is
+initialized only once and thus no shared resource is released until the last
+finalization call is made.
+}
+
+
+\subsection{Dynamic Process Creation}
+\label{subsec:interoperability:mpmd}
+
+MPI defines a dynamic process model that allows creation of processes after
+an MPI application has started (e.g., by calling \FUNC{MPI\_Comm\_spawn}) and
+connection to independent processes (e.g., through \FUNC{MPI\_Comm\_accept}
+and \FUNC{MPI\_Comm\_connect})
+and provides a mechanism to establish communication
+between the newly created processes and the existing MPI application (see
+MPI standard version 3.1, Chapter 10).
+Unlike MPI, \openshmem starts all processes at once and requires all PEs to
+collectively allocate and initialize resources (e.g., symmetric heap) used by
+the \openshmem library before any other \openshmem routine may
+be called. Communicating with a dynamically created process in the \openshmem
+environment may result in undefined behavior.
+Hence, users should not use \openshmem and MPI dynamic process models
+in the same program.
```

Comment:
I'm not sure this is the right conclusion.  Connect/accept can be used to connect a job to a service, e.g. connect a SHMEM job to a database or I/O service.  Or to connect two different SHMEM jobs as a workflow.  These should all work fine, provided the communication is done with MPI and not SHMEM.  A better conclusion may be:
```
\openshmem does not support communication with dynamically created or connected processes. 
In such a scenario, MPI must be used communicate with these processes.
```

Reply:
I like the suggested conclusion. I disallowed it because I think it is hard to precisely cover all combinations of `SHMEM+MPI dynamic process` (see below). 

- case 1:
After MPI initialization and before any PEs start OpenSHMEM initialization, the program spawns(connects to) a new MPI process. It is unclear whether P1 is included in the original `TEAM_WORLD`.
```
P0                  Spawned P1
MPI_Init(..)
spawn();
shmem_init();       shmem_init();
```

- case 2:
After MPI and SHMEM initialization, the program spawns(connects to) a new process and the new process wants to communicate with PEs in the original `TEAM_WORLD`.
```
P0                  Spawned P1
MPI_Init(..)
shmem_init();
spawn();
                    shmem_init();
```

- case 3:
After MPI and SHMEM initialization, the program spawns(connects to) some new processes (e.g., P1 and P2 in the example). Can `P1` and `P2` communicate with each other through OpenSHMEM? 
```
P0                  Spawned P1      Spawned P2
MPI_Init(..)
shmem_init();
spawn();
                    shmem_init();   shmem_init();
```

I think the new conclusion clearly defines all the above cases!